adr/0026-schema.md

```diff
@@ -0,0 +1,67 @@
+
+- `setVariable` deprecated in favor of `setVariables`
+- `scripts` deprecated in favor of `actions`
+- `required` soon to be deprecated in favor of `optional`.
+- `group` deprecated in favor of `flavor`, however these work functionally different and cannot be automatically migrated
+- `cosignKeyPath` deprecated in favor of specifying the key path at create time
+
+Zarf has not disabled any deprecated keys thus far. On create the user is always warned when using a deprecated field, however the field still exists in the schema and functions properly. Some of the deprecated keys can be migrated automatically as the old item is a subset or directly related to it's replacement. For example, setVariable is automatically migrated to a single item list of setVariables. This migration occurs on the zarf.yaml fed into the package during package create, however the original field is not deleted from the packaged zarf.yaml because the Zarf binary used in the airgap or delivery environment is not assumed to have the new schema fields that the deprecated key was migrated to.
+
+There are two problems this ADR aims to solve
+- Supporting deprecated features forever is complex and time consuming
+- Dropping feature support can frustrate or confuse users
+
+# Decisions
+
+Zarf will introduce proper schema versions. A top level key, `apiVersion`, will be introduced to allow users to specify the schema. If `apiVersion` is not specified we will break on deploy and instruct users to run `zarf dev update-schema`. Any package without this key will be assumed to use the alpha schema
+
+In v1, deprecated features that have a direct migration path will cause an error on create, but will still be deployed if the package was created pre v1. If a feature does not have a direct automatic migration path (cosignKeyPath & groups) they will fail on deploy. This will exist until the alpha schema is entirely removed from Zarf, which will happen one year after v1 is released.
+
+The existing types which comprise the Zarf schema will be moved to types/alpha and be renamed to alpha. Once v1 is released this package will not be touched again. A copy of these types will be created in a package called v1
+
+Any key that exists at the introduction of v1 will last the entirety of that schema lifetime. The features may be deprecated, but will not be removed until the next schema version.
+
+The `zarf dev update-schema` command will be introduced to automatically update deprecated fields in the users `zarf.yaml` where possible. It will also add the apiVersion key and set it to v1. `zarf package create` will fail if the user has deprecated keys with a list of the deprecated keys and a call to action to run `zarf dev update-schema`.
+
+
+## BDD scenarios
+### 1 create with deprecated keys
```

Comment:
I might be misunderstanding something here, but it seems like the `create with deprecated keys` scenario contradicts the `pre 1 auto migrate create -> 1 deploy` scenario.

The former says this about running `zarf p create` w/ deprecated keys pre v1:

>
- *Given*: A user has a `zarf.yaml` with keys deprecated pre v1
- *when* the user runs `zarf package create`
- *then* they will receive an error and be told to run `zarf dev update-schema` or how to migrate off cosign key paths or how to use flavors over groups depending on the error

The latter says this about running `zarf p create` w/ deprecated keys pre v1:

>  
- *Given*: A package is created with Zarf pre v1
- *and* that package has deprecated keys that can be automatically migrated (required, scripts, & set variables)
- *when* the package is deployed with Zarf v1
- *then* the keys will be automatically migrated & the package will be deployed without error.

Having `zarf package create` error out on deprecated keys might be frustrating for users on pre v1 version. It’s better to give a warning with suggestions so they can update without being completely blocked from creating packages.

Also, I believe the automatic migrations would happen during `package create` rather than `package deploy`. This is what we currently do in `ComposeComponents`

https://github.com/defenseunicorns/zarf/blob/fd452917584382f7ed588ffc5eaf428ecc7bb868/src/pkg/packager/creator/compose.go#L43

Reply:
They weren't worded well, but they don't contradict as the first situation involves creating a package on Zarf v1 whereas the latter is a packaged created pre v1 that is deployed on v1. We want the existing packages to still work on deploy, but we don't want new packages to be made using the deprecated fields.  Zarf will have a `zarf dev update-schema` command that is going to do everything automatically for users except for features that are removed (groups and cosignKeyPath). In addition, we want to force users to add an apiVersion. This allows us to assume that any package without an apiVersion is alpha (the current schema).

You're correct in that we currently do the automatic migration at create time, I just wanted to be explicit that packages using deprecated keys with automatic migration paths will still be supported in v1. The only situation this can happen is if the package was built pre v1.